core/server/worker/src/main/java/alluxio/worker/block/BlockWorkerClientServiceHandler.java

```diff
@@ -147,18 +150,36 @@ public String toString() {
    * @throws AlluxioTException if an Alluxio error occurs
    */
   @Override
-  public LockBlockResult lockBlock(final long blockId, final long sessionId)
-      throws AlluxioTException {
+  public LockBlockResult lockBlock(final long blockId, final long sessionId,
+      final LockBlockTOptions options) throws AlluxioTException {
     return RpcUtils.callAndLog(LOG, new RpcCallable<LockBlockResult>() {
       @Override
       public LockBlockResult call() throws AlluxioException {
-        long lockId = mWorker.lockBlock(sessionId, blockId);
-        return new LockBlockResult(lockId, mWorker.readBlock(sessionId, blockId, lockId));
+        if (!options.isSetUfsPath() || options.getUfsPath().isEmpty()) {
+          long lockId = mWorker.lockBlock(sessionId, blockId);
+          Preconditions.checkState(BlockLockIdUtils.isAlluxioBlockLockId(lockId));
+          return new LockBlockResult(lockId, mWorker.readBlock(sessionId, blockId, lockId));
+        }
+
+        long lockId = mWorker.lockBlockNoException(sessionId, blockId);
+        if (BlockLockIdUtils.isAlluxioBlockLockId(lockId)) {
+          return new LockBlockResult(lockId, mWorker.readBlock(sessionId, blockId, lockId));
+        }
+        // When the block does not exist in Alluxio but exists in UFS, try to open the UFS
+        // block.
+        if (mWorker.openUfsBlock(sessionId, blockId, new OpenUfsBlockOptions(options))) {
+          lockId = BlockLockIdUtils.UFS_BLOCK_LOCK_ID;
```

Comment:
possible not to overload `lockId` but add a new field in `LockBlockResult` to indicate the lock type / status. Feel free to leave this as a TODO. Essentially, I prefer not overloading the lockId to indicate different meanings.